### ISSUE TITLE
[Framework] Should we split the library up into multiple NuGet packages?

### ISSUE THREAD

**Issue by craftworkgames** [craftworkgames/MonoGame.Extended#275]
The idea of splitting the library up into multiple packages has come up a few times. Most recently in PR #274. I've extracted a few of the quotes from that discussion as a starting point.

> Splitting the system up into projects means you now need a new nuget package for each project (that seems like a PITA) and even if you didn't, why would you need this?
> 
> Do we want a "core" library? 
> 
> Or is it acceptable to have multiple common feature libraries such as TextureAtlases or even Primitives (when it gets flushed out more) which most people would most likely have to add as dependencies separately?
> 
> Also while we are on the subject of moving around code, perhaps we could add the namespace for 2D and 3D?
> 
> While this would imply major breaking changes, this would be beneficial in the long term. A loosely coupled architecture would be very flexible.
> 
> What would work in many cases would be add some well designed interfaces in the core library, that would decouple things. But this would imply more work to the user for instantiation. A IoC/dependency injection solution would then be needed.

**Comment by LithiumToast** [comment]
One pro to splitting up the codebase into multiple library projects is that the code can be released as grouped modules. Currently there has not been a release in quite sometime, so, perhaps this would solve that problem and allow for code to be released faster.

**Comment by raizam** [comment]
Also as a first step toward modularity, we could start defining interfaces on the core lib. 
But we would need to examine for each case if this is really relevant, to make sure it doesn't get overkill.
As a DI library I have used many, and the best one by far to me is LightInject.
Maybe the creation of an 'experimental' branch would be useful.

About Portable classes libraries, I think the current setup (Windows 8/Silverlight ect..) will soon be obsolete, as the .NETStandard version will replace it. The reason I needed it is that currently the portable class API doesn't doesn't include the Thread/ThreadPool classes, and I wanted to contribute in this area.

**Comment by LithiumToast** [comment]
I don't think dependency injection is necessary at all here. All the code is already cross-platform by virtue of using PCLs right now, so using dependency injection for platform specific is not appropriate here. Also, dependency injection can get quite complex and use reflection depending on what library you use. XNA / MonoGame already have a service container on the `Game` class which is essentially just a `Dictionary<Type, object>` used to lookup services by type. If we absolutely need inversion of control we can use that instead of NInject, Autofac, LightInject or some other IoC / DI library.

> we could start defining interfaces on the core lib. 

I don't think most or any of code needs abstraction?

The dependencies are pretty much just that feature X, requires feature Y. So to use library A with feature X, the library B with feature Y must be added as a reference. This isn't really a big deal as long as the libraries have good names.

**Comment by craftworkgames** [comment]
> One pro to splitting up the codebase into multiple library projects is that the code can be released as grouped modules. Currently there has not been a release in quite sometime, so, perhaps this would solve that problem and allow for code to be released faster in the future.

Yes, absolutely.

So long as the overhead of having multiple packages isn't a worse problem. The good news is, I'm currently working getting a new builder server setup and I plan to automate package building.

> Also as a first step toward modularity, we could start defining interfaces on the core lib. 
But we would need to examine for each case if this is really relevant, to make sure it doesn't get overkill.

I think the first step is figuring out *why* we want to split the library up?  Once we know the *why* the *how* becomes a lot more obvious.

>About Portable classes libraries, I think the current setup (Windows 8/Silverlight ect..) will soon be obsolete, as the .NETStandard version will replace it.

I raised issue #262 a while ago to start thinking about .NET Standard. I think the big question mark here is how it fits with MonoGame. My feeling is that it's probably way more difficult than it sounds.

> Also, dependency injection can get quite complex and use reflection depending on what library you use. XNA / MonoGame already have a service container on the Game class which is essentially just a Dictionary<Type, object> used to lookup services by type. If we absolutely need inversion of control we can use that with just plain constructor injection instead of NInject, Autofac, LightInject or some other IoC / DI library.

Agreed. I don't think we need to introduce any IoC containers into the library. We certainly don't *need* one, and I'm struggling to see the benefit of *having* one so I certainly don't want to force anyone to use one if they don't have too.

> Also while we are on the subject of moving around code, perhaps we could add the namespace for 2D and 3D?

I'm not entirely certain how or when we're going to deal with 3D stuff. However, I do think it's worth considering how 3D fits into the package discussion. It would be nice to think that if you're making a 3D game the set of packages you're installing would be quite different to a 2D game. The reality is going to be a little more grey I suspect. Especially when it comes to things like the `Camera`. Of course, we're not the first ones to solve this problem, we don't have to look far to figure out how other libraries do it.

For me, the primary reason to even consider splitting the library up is to give users some choices. For example, if they don't want to use Tiled in their game they don't install the `MonoGame.Extended.Tiled` package. Or perhaps they want to use an alternative GUI library, or collision detection, etc.

This probably means the `MonoGame.Extended` package becomes the "core" package with a bunch of common dependencies like texture regions / atlases, sprites, fonts and some interfaces perhaps. Then the more "major" features become separate packages.

The tricky part is defining exactly what goes in "core" and what belongs in a separate package. I am curious to know at a high level what people think belongs in "core". However, we could spend weeks arguing about the details so at some point I'm going to have to make a final decision.

Another approach (probably a more sensible one) would be to pick *one* feature that is an obvious candidate for being split out and start there. For example, it's already clear that there could many different GUI packages, so maybe that's the starting point.

**Comment by bsimser** [comment]
I'm on-board with splitting the packages up where it makes sense (I just felt in the PR it was simply done by namespace which resulted in far too many packages).

I think the key thing you mentioned was a) making `MonoGame.Extended` the core package and b) breaking up the features into sensible dependent libraries (like Tiled, or a GUI, etc.).

As for what should be in core we should just take a look at what we think is core and propose a straw dog structure for what would be included. Maybe going backwards makes sense (and the tiled is probably a key one) and might help flesh out what is in core. If you think of the core system as being what everyone depends on and it includes things that any game would use that should help form the basis for the split.

Looking forward to this and we won't get it right but I always believe version 1, it's not perfect, but ship it anyway.

**Comment by craftworkgames** [comment]
Posting this here from gitter so we don't loose it.

> Lucas Girouard-Stranks @LithiumToast 12:35
> Animations, Tiled, Collisions, Entities, Graphics (the stuff I wrote I want to move it to a rendering library), > Gui should probably all be seperate libraries
> oh particles too
> what is left is BitmapFonts, ViewportAdapaters, Timers, TextureAtlases, Serialization Helpers, Primitives, > Collections, Cameras which would be in the core

**Comment by craftworkgames** [comment]
I've created a new branch called [framework-split](https://github.com/craftworkgames/MonoGame.Extended/tree/framework-split) to experiment with how this is going to work.

The first thing I tried was moving the `Collisions` code into a separate project because it's a small but crappy bit of the library. I've wanted to remove it from the library for a while, but I'm not sure how many people are actually using it. Moving it out of the core mitigates this problem a little.

For the most part, it's pretty easy. I just created a new project that matches the namespace of the code I want to move and update the references. I haven't worked out how this is going to work with NuGet yet. That's next on the to-do list.

One problem I did encounter was the `CollisionWorldExtensions` class. This depends on both the `Collisions` and the `Tiled` code. For now I've just moved it into the `Collisions` project, but this won't hold once the `Tiled` code is moved into it's own project.

The *proper* fix is to create appropriate interfaces in the core library and have both the `Collisions` and `Tiled` projects know about just the interfaces. It would properly decouple the code and *should* result in a better design anyway.

So far I'm pretty happy with this approach. It'll be interesting to see how it plays out as NuGet packages.
